### PR TITLE
fix(vault): resource / key parsing problematic with some vaults

### PR DESCRIPTION
**Pull request by bungle**
### Summary

The vault has generic part, implemented in `kong/pdk/vault.lua` and then the implementation strategy part implemented in `kong/vaults/*`.

The interface between them is:
```
string, err, ttl = get(conf, resource, version)
```

The generic part supported also a sub-key lookup in case the vault `resource` returned a multi-value answer as a JSON string. An example of this:

```
❯ PG='{"username":"john","password":"qwerty"}' kong vault get env/pg
{"username":"john","password":"qwerty"}
❯ PG='{"username":"john","password":"qwerty"}' ./bin/kong vault get env/pg/username
john
```

In both examples the `resource` given to Vault `strategy` is `test`. See how we used the last slash to determine where comes the multi-value lookup key. For env vault this works fine, but it is problematic with other vaults as we cannot determine whether `pg/username` is the name of the actual secret or is the `pg` the name of actual secret, and when it is the first (`pg/username`), we will endup giving just `pg` to Vault strategy.

This PR changes the character used for multi-value lookups from that `/` to `:`, so example goes like this:

```
❯ PG='{"username":"john","password":"qwerty"}' kong vault get env/pg
{"username":"john","password":"qwerty"}
❯ PG='{"username":"john","password":"qwerty"}' ./bin/kong vault get env/pg:username
john
```

In AWS Vault the secret name can be:
- `my/secret`, or full ARN like:
- `arn:aws:secretsmanager:us-east-2:123456123456:secret/my/secret-ABCDEFG`

The first form does not allow colons `:`, but the second is filled with colons.

So the key detection function starts from the end of string. If it finds `/`, it will stop and treat the secret as a single value secret. If it finds `:` (before finding `/`), it takes the left part from that as a `resource` and the right part as a `key`. Thus in:

- `my/secret` resource will be `my/secret` and the key is `nil`
- `arn:aws:secretsmanager:us-east-2:123456123456:secret/my/secret-ABCDEFG` the resource will be `arn:aws:secretsmanager:us-east-2:123456123456:secret/my/secret-ABCDEFG` and the key is `nil`
- `my/secret/another-path:key` resource will be `my/secret/another-path` and the key is `key`
- `arn:aws:secretsmanager:us-east-2:123456123456:secret/my/secret-ABCDEFG:key` the resource will be `arn:aws:secretsmanager:us-east-2:123456123456:secret/my/secret-ABCDEFG` and the key is `key`.

This is a breaking change.

KAG-1372

See also: https://github.com/Kong/docs.konghq.com/pull/5501

### CODE REVIEW

**Comment by bungle** [comment]
@hbagdi this is something I wish you to give some input as it is a breaking change. I am not sure do we need to go through all the entities that could possibly hold secrets and migrate them. Certainly we cannot go through all the `kong.conf` settings. On the other hand if someone tried to use secret name like `my/secret` in AWS he would have failed miserably already, which kinda speaks that migration is not necessary. I will surely add a better test suites for the `parse_reference` function to cover this, but I would like to get answer to this change if we can get it in. It causes currently problems with almost all the Vault implementations (AWS, GCP, Azure).

**Comment by bungle** [comment]
This will also need update:
https://docs.konghq.com/gateway/latest/kong-enterprise/secrets-management/reference-format/

**Comment by hbagdi** [comment]
Such are the pains of shoving in structured data into a string representation that is not extensible at all.

Anyway,
1. What are the downsides of continuing to use `/` for env vault to minimize the breaking change?
2. This is not a breaking change for AWS users. Is that understanding accurate?
3. What is the impact of Hashicorp Vault usage here?
4. What is the imapct of Google cloud?

We need to be sure about these impacts and only then we can make a decision.
Given the adoption, I'm fine with making such a change if the impact on other Vault is same as AWS - nearly non-existent.

**Comment by bungle** [comment]
> Such are the pains of shoving in structured data into a string representation that is not extensible at all.

And url is very much extensible and has structure. We made mistake of using same separator for resource and key that we didn't just see in original (that is we didn't anticipate the fact that vault keys can have `/`. Very normal to miss these details, and very very easy to point after it has brought to attention), and now we are fixing it by adding more structure. If we went with that always add a structured field on side of every field where we could store a secret, then our schemas would have exploded in fields, and I would say that much much worse things would have happened. The reference format was probably the best invention of secrets management. But let's leave those notes aside from here.
 
> 1. What are the downsides of continuing to use `/` for env vault to minimize the breaking change?

It would leak implementation (strategy) specific details to generic side of code. I think it is rather uncommon to have multivalue env vars, so it is rarely used there.

> 2. This is not a breaking change for AWS users. Is that understanding accurate?

If AWS has secret of /dog/cat and that is multi value then /dog/cat/key would work correctly.  The problem comes when /dog/cat/key is a single value, in which case we will send /dog/cat to AWS strategy and it cannot find the secret as we gave a partial path. So on the first case (that works correctly currently) it would be a breaking.

> 3. What is the impact of Hashicorp Vault usage here?

I think more or less the same if hashicorp allows `/` in keys, but afaik it does not (not 100% sure though).

> 4. What is the imapct of Google cloud?

This was brought to attention when implementing the Azure Vault, but I think GCP probably has same issue, but it may have resolved with hack. This PR is to get rid of such hacks, such as this is not needed: https://github.com/Kong/kong-ee/pull/4406/files#diff-8ace3004fb2152fbb7b218aa0024086aa145c6132f6b223cb1d48afccd3decd4R27-R38 (example from Azure vault).

> We need to be sure about these impacts and only then we can make a decision. Given the adoption, I'm fine with making such a change if the impact on other Vault is same as AWS - nearly non-existent.

Yes. New big features always have some child defects in their first incarnations. Especially when it comes to integration in a space where there in no standardization, like Vaults.

**Comment by bungle** [comment]
The related docs pr is here:
https://github.com/Kong/docs.konghq.com/pull/5501

**Comment by hbagdi** [comment]
> very easy to point after it has brought to attention

True. I was wondering that out loud, my intentions are not to point a finger.


> This was brought to attention when implementing the Azure Vault, but I think GCP probably has same issue, but it may have resolved with hack.

Let's not resolve it with a hack. We are in (relatively) early days of this feature and we should break if that is the right thing to do at this point.

**Comment by hbagdi** [comment]
Do we do arbitrary level of nesting for multi-value secrets?
If not, can this be done in a non-breaking fashion where the last path element is looked up as a key within an object as well as the secret name (in some precedence order)? We can still introduce colon as a separator and recommend only that. The 'hack' I suggest here would be done only to avoid a breaking change.
I'm not sure if this leads to bigger problems or not. Looking forward to your response.

**Comment by bungle** [comment]
> Do we do arbitrary level of nesting for multi-value secrets?
> If not, can this be done in a non-breaking fashion where the last path element is looked up as a key within an object as well as the secret name (in some precedence order)?

We do support multiple level of nesting, and currently look the last slash.

**Comment by teb510** [comment]
this makes sense to me. suggest we go ahead. 

does this change also mean that no 'hacky changes' are required to make Azure vault function in the future?

i think this feature adds sufficient value that a small breaking change here is ok.

**Comment by bungle** [comment]
> does this change also mean that no 'hacky changes' are required to make Azure vault function in the future?

I think it solves the issue @tysoekong was having with Azure.

**Comment by bungle** [comment]
Azure says:
> Name: Type a name for the secret. The secret name must be unique within a Key Vault. The name must be a 1-127 character string, starting with a letter and containing only 0-9, a-z, A-Z, and -. For more information on naming, see [Key Vault objects, identifiers, and versioning](https://learn.microsoft.com/en-us/azure/key-vault/general/about-keys-secrets certificates#objects-identifiers-and-versioning)

**Comment by bungle** [comment]
I think in Hashicorp Vault the secret can contain any chars:

```
❯ vault kv put -mount secret "t/es:t" testv="h1:v1"
=== Secret Path ===
secret/data/t/es:t
```

**Comment by bungle** [comment]
AWS says:
> The secret name can contain ASCII letters, numbers, and the following characters: /_+=.@-
-- https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_CreateSecret.html#SecretsManager-CreateSecret-request-Name

Which means that the `:` is not a problem there.

**Comment by hbagdi** [comment]
How about `?`?
There maybe other vaults that we expand usage for and we should go for something that is unlikely to cause an issue.

**Comment by bungle** [comment]
> How about `?`? There maybe other vaults that we expand usage for and we should go for something that is unlikely to cause an issue.

Our format is URL so `?` is for query string config overrides like:
```
{vault://hcv/pg-password?ttl=20}
```

**Comment by hbagdi** [comment]
> we can document that people should not use : in secret names

No way to escape the `:`. I know that opens up another can of worms.
This will be fine for most parts, I'm trying to be as accommodating as possible.
I leave the decision up to you.

File: spec/02-integration/13-vaults/02-env_vault_spec.lua [hunk 1]
```diff
@@ -55,12 +55,11 @@ describe("Environment Variables Vault", function()
     finally(function()
       helpers.unsetenv("TEST_ENV_JSON")
     end)
-    local res, err = get("{vault://env/test_env_json/username}")
+    local res, err = get("{vault://env/test_env_json:username}")
```
Comment (jschmid1) [inline spec/02-integration/13-vaults/02-env_vault_spec.lua:58]: This sounds like a breaking change to me. We have to offer some sort of migration path.

**Comment by jschmid1** [comment]
I'm on board with the fundamentals of this change. I think there isn't a single character that is forbidden in all current (and whatever the future brings) vault implementations. This means we have to bite the bullet and agree on something. `:` sounds fine to me.

Something I'll give to consider though. We currently only allow to target keys in _one_ level of nesting. 


```
{vault://hcv/foo/bar}

# returns {"username": "my-user", "password": "my-pass"}

{vault://hcv/foo/bar:username}
# returns "my-user"
```

A more flexible, yet more convoluted option would be (ignore the choice of syntax, can be any of `$(), ->, ${}, |.|.|`)
```
{vault://hcv/foo/nesting}

# returns 
{
  "parent": {
    "username": "my-user",
    "password": "my-pass"
  }
}

{vault://hcv/foo/bar$(parent.username)}
# returns "my-user"

# the example from before would turn from 

{vault://hcv/foo/bar:username}
# returns "my-user"

to 

{vault://hcv/foo/bar$(username)}

```

**Comment by bungle** [comment]
@jschmid1,

I checked the thing again,

We might not need this after all. See: https://github.com/Kong/kong/blob/master/kong/pdk/vault.lua#L373-L375

The logic goes like this:
```
{vault://hcv/foo/bar}
```

The `hcv` gets `foo` and `bar` is used as a `key` (multivalue).

```
{vault://hcv/foo/bar/}
```

The `hcv` gets `foo/bar` and there is no key (single value).

It may not be the most natural, but it works, and no changes or migrations are required and this PR can be closed?

Did I miss something?

ping @hbagdi, @tysoekong

**Comment by jschmid1** [comment]
Huh, ending the reference with a slash to indicate a single-value that does **not** need to get accessed further isn't that bad.
It isn't very pleasing to the eye but if it spares us breaking changes and possible migrations I'm all up for it.

**Comment by kikito** [comment]
Closing this for now, we can use / at the end as a workaround

**Pull request closed without merging**